pkg/etcdcli/health.go

```diff
@@ -41,66 +42,34 @@ type healthCheck struct {
 type memberHealth []healthCheck
 
 func getMemberHealth(ctx context.Context, etcdMembers []*etcdserverpb.Member) memberHealth {
-	var wg sync.WaitGroup
 	memberHealth := memberHealth{}
-	hch := make(chan healthCheck, len(etcdMembers))
 	for _, member := range etcdMembers {
 		if !HasStarted(member) {
 			memberHealth = append(memberHealth, healthCheck{Member: member, Healthy: false})
 			continue
 		}
-		wg.Add(1)
-		go func(member *etcdserverpb.Member) {
-			defer wg.Done()
-			// If the endpoint is for a learner member then we should skip testing the connection
-			// via the member list call as learners don't support that.
-			// The learner's connection would get tested in the health check below
-			skipConnectionTest := false
-			if member.IsLearner {
-				skipConnectionTest = true
-			}
-			cli, err := newEtcdClientWithClientOpts([]string{member.ClientURLs[0]}, skipConnectionTest)
-			if err != nil {
-				hch <- healthCheck{Member: member, Healthy: false, Error: fmt.Errorf("create client failure: %w", err)}
-				return
-			}
-			defer func() {
-				if err := cli.Close(); err != nil {
-					klog.Errorf("error closing etcd client for getMemberHealth: %v", err)
-				}
-			}()
-
-			st := time.Now()
-			ctx, cancel := context.WithCancel(ctx)
-			var resp *clientv3.GetResponse
-			if member.IsLearner {
-				// Learner members only support serializable (without consensus) read requests
-				resp, err = cli.Get(ctx, "health", clientv3.WithSerializable())
-			} else {
-				// Linearized request to verify health of a voting member
-				resp, err = cli.Get(ctx, "health")
-			}
 
-			hc := healthCheck{Member: member, Healthy: false, Took: time.Since(st).String()}
-			if err == nil {
-				if resp.Header != nil {
-					raftTerms.Set(member.Name, resp.Header.RaftTerm)
-				}
-				hc.Healthy = true
-			} else {
-				klog.V(2).Infof("health check for memberID (%v) failed: err(%v)", resp.Header.MemberId, err)
-				hc.Error = fmt.Errorf("health check failed: %w", err)
-			}
-			cancel()
-			hch <- hc
-		}(member)
-	}
+		const defaultTimeout = 30 * time.Second
+		resChan := make(chan healthCheck, 1)
+		go func() {
+			ctx, cancel := context.WithTimeout(ctx, defaultTimeout)
+			defer cancel()
+
+			resChan <- checkSingleMemberHealth(ctx, member)
+		}()
 
-	wg.Wait()
-	close(hch)
+		select {
```

Comment:
@tjungblu this select will be running within the loop, so it will block till the go routine write to the channel, or timeout. IMO it would be more concurrent if you spawn the routines from the loop, and after the loop you gather the result (i.e. the original impl using sync.WG) 

This way, you block within each iteration waiting for the result or timeout .. please correct me if I am wrong

Reply:
yeah that was the actual intention of removing it, I don't really understand what you gain from parallelizing three at max 10ms calls here on the happy path (30ms vs. 10ms?). Timeout sucks as it has a 30s penalty, but that is the same when you're parallelizing it.

Reply:
gotcha, ok make sense then 👍🏽